lib/backend/k8s/resources/resources.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resources
+
+import (
+	"encoding/json"
+
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+)
+
+const (
+	labelsAnnotation      = "projectcalico.org/Labels"
+	annotationsAnnotation = "projectcalico.org/Annotations"
+)
+
+// Interface that all Kubernetes and Calico resources implement.
+type Resource interface {
+	runtime.Object
+	metav1.ObjectMetaAccessor
+}
+
+// Interface that all Kubernetes and Calico resource lists implement.
+type ResourceList interface {
+	runtime.Object
+	metav1.ListMetaAccessor
+}
+
+type ConvertK8sResourceToKVPair func(Resource) (*model.KVPair, error)
+
+// Store Calico Metadata in the k8s resource annotations for non-CRD backed resources.
+// Currently this just stores Annotations and Labels and drops all other metadata
+// attributes.
+func SetK8sAnnotationsFromCalicoMetadata(k8sRes Resource, calicoRes Resource) {
+	a := k8sRes.GetObjectMeta().GetAnnotations()
+	if a == nil {
+		a = make(map[string]string)
+	}
+	if labels := calicoRes.GetObjectMeta().GetLabels(); len(labels) > 0 {
+		if lann, err := json.Marshal(labels); err == nil {
+			a[labelsAnnotation] = string(lann)
+		}
+	}
+	if annotations := calicoRes.GetObjectMeta().GetAnnotations(); len(annotations) > 0 {
+		if aann, err := json.Marshal(annotations); err == nil {
```

Comment:
We shouldn't be throwing away the marshalling error here - we need to log it out at least.

Above as well